chore(client): Migration of SolutionViewer component to Typescript.

ashiskumar-1999 opened the pull request:
<!-- Please note that low quality PRs and PRs that do not follow our contributing guidelines will be marked as invalid for Hacktoberfest. -->

Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Reference: #42256

<!-- Feel free to add any additional description of changes below this line -->

ashiskumar-1999 committed:
The Solutionviewer .js file has been migrated into the .tsx file.

ieahleen commented:
Hey @ashiskumar-1999 

We would love to be able to merge your changes but it looks like there is an error with the CI build. ⚠️

Once you resolve these issues, we will be able to review your PR and merge it. 😊

---

Feel free to reference the [contributing guidelines](https://contribute.freecodecamp.org/#/how-to-work-on-coding-challenges?id=testing-challenges) for instructions on running the CI build locally. ✅

ashiskumar-1999 commented:
Hey @ieahleen, I have run the test command locally and it passed without any errors. Can you tell me specifically about the issues that you have mentioned.

ieahleen commented:
Sorry, it failed on files you did not change

nhcarrigan commented:
@ieahleen @ashiskumar-1999 the errors in the checks are TypeScript compile errors. Because the `ProjectModal` was converted to TypeScript, the `show-project-links` is throwing compile errors due to the new ProjectModal type definitions not matching.

Give me a few moments and I'll propose some fixes. 🙂

nhcarrigan commented on client/src/components/SolutionViewer/ProjectModal.tsx:
```diff
@@ -54,7 +53,6 @@ const ProjectModal = props => {
   );
```
I can't reach the line because of GitHub restrictions, but above here on line 47:

Replace
```tsx
        <SolutionViewer challengeFiles={challengeFiles} solution={solution} />
```
with
```tsx
        <SolutionViewer
          challengeFiles={challengeFiles}
          solution={String(solution)}
        />
```

Going to cc @ojeytonwilliams on this - it's a cheaty approach to address the type def, but might be better to re-evaluate the type def?

nhcarrigan replied:
![image](https://user-images.githubusercontent.com/63889819/136032414-1c17f054-22e8-4bd1-9f3c-0bf13878effc.png)

Looks like this hasn't been changed yet.

nhcarrigan commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -0,0 +1,67 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+import { Panel } from '@freecodecamp/react-bootstrap';
+import Prism from 'prismjs';
+import React from 'react';
```
```suggestion
import React from 'react';
import { ChallengeFile, ChallengeFiles } from '../../redux/prop-types';
```
Necessary import

nhcarrigan commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+
+interface solutionviewerProps {
+  challengeFiles: [];
```
```suggestion
  challengeFiles: ChallengeFiles;
```
Proper TypeDef

nhcarrigan commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -0,0 +1,67 @@
+  return (
+    <>
+      {challengeFiles?.length ? (
+        challengeFiles.map((challengeFile: string) => (
```
```suggestion
        challengeFiles.map((challengeFile: ChallengeFile) => (
```
This is not a string.

nhcarrigan commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -0,0 +1,67 @@
+                <code
+                  dangerouslySetInnerHTML={{
+                    __html: Prism.highlight(
+                      challengeFile.trim(),
+                      Prism.languages[challengeFile],
```
```suggestion
                      challengeFile.contents.trim(),
                      Prism.languages[challengeFile.ext],
```
Correct property access.

nhcarrigan commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -0,0 +1,67 @@
+            className='solution-viewer'
+            key={challengeFile}
+          >
+            <Panel.Heading>{challengeFile.toUpperCase()}</Panel.Heading>
```
```suggestion
            <Panel.Heading>{challengeFile.ext.toUpperCase()}</Panel.Heading>
```
Correct property access.

nhcarrigan commented on client/src/components/SolutionViewer/ProjectModal.tsx:
```diff
@@ -1,11 +1,11 @@
-  challengeFiles: PropTypes.array,
+type projectmodalProps = {
+  challengeFiles: [];
   // TODO: removed once refactored to TS
```
This comment can be removed now.

nhcarrigan commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -0,0 +1,67 @@
+import Prism from 'prismjs';
+import React from 'react';
+
+interface solutionviewerProps {
```
```suggestion
type SolutionViewerProps = {
```
PascalCase, also don't believe this needs to be an interface.

nhcarrigan commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -0,0 +1,67 @@
+function SolutionViewer({
+  challengeFiles,
+  solution = '// The solution is not available for this project'
+}: solutionviewerProps): JSX.Element {
```
```suggestion
}: SolutionViewerProps): JSX.Element {
```
PascalCase

naomi-lgbt reviewed (changes requested):
These changes should resolve the TypeScript compile errors (unless I missed one), and allow us to move forward with the review process. 🙂

ashiskumar-1999 commented:
Hey @nhcarrigan, Have you done the changes in the pull request or just suggest doing it?
Because when I see the commits, your changes are in the first. So, I am confused a little bit.

ojeytonwilliams commented:
@ashiskumar-1999 they're just suggestions. If you look at https://github.com/freeCodeCamp/freeCodeCamp/pull/43694/commits you can see there's only your original commit.

ashiskumar-1999 commented:
Thank you @ojeytonwilliams, I am changing the code accordingly and push it to the pull request.

ieahleen commented:
@ashiskumar-1999 you can commit from the suggestions, go to /files, then use the button "Add suggestion to batch", and once you have finished at the top there is a "Commit" button that will let you commit all the suggestions

ashiskumar-1999 committed:
Apply suggestions from code review

I have added the suggestions and before adding them to the commit, I have checked it once locally and it doesn't throw any new error.

Co-authored-by: Nicholas Carrigan (he/him) <nhcarrigan@gmail.com>

ashiskumar-1999 commented:
Hey @ieahleen @ojeytonwilliams, I have made the changes in my commits. You can now review it and merge it if there will be no more errors.

ashiskumar-1999 committed:
Update the SolutionViewer.tsx 

Added the "null" and "undefined" type to "solution" property.

ashiskumar-1999 commented:
Hey @nhcarrigan, I have added all the suggestions but this error came up. 
Error: client/src/components/SolutionViewer/SolutionViewer.tsx(46,16): error TS2531: Object is possibly 'null'.
Can you help me to identify this issue?

nhcarrigan commented on client/src/components/SolutionViewer/SolutionViewer.tsx:
```diff
@@ -0,0 +1,68 @@
+
+type SolutionViewerProps = {
+  challengeFiles: ChallengeFiles;
+  solution: string | null | undefined;
```
```suggestion
  solution: string;
```
This one should stay a string - when you handle the fix in the other comment, the error for this being possibly null/undefined will go away.

naomi-lgbt reviewed (changes requested):
Couple of steps should fix that error.

ashiskumar-1999 committed:
Update SolutionViewer.tsx

ashiskumar-1999 committed:
Update ProjectModal.tsx

ashiskumar-1999 committed:
Update ProjectModal.tsx

ashiskumar-1999 commented:
Hey @ieahleen, Can you now review it?
I think there is no error left to be resolved but there are two checks pending.

gikf commented:
> I think there is no error left to be resolved but there are two checks pending.

I've poked them for you :slightly_smiling_face:

ashiskumar-1999 commented:
Hey @gikf, Can we ignore this prettier error?

gikf commented:
> Hey @gikf, Can we ignore this prettier error? When I did some changes yesterday, the SolutionViewer.tsx had no prettier issue but now it came up. You can see it above.

It will need to be changed. Looks like on that passing commit only labelling CI was run - see https://github.com/freeCodeCamp/freeCodeCamp/runs/3804026331, hence no linting error then.

ashiskumar-1999 commented:
So, Can you tell me what to do so that I can avoid this prettier issue?

ashiskumar-1999 commented:
I have cross-checked every line of code. It seems like there will be no ";" need in the code but I don't know why this error has come?

ashiskumar-1999 committed:
Update SolutionViewer.tsx

ShaunSHamilton commented on client/src/components/SolutionViewer/ProjectModal.tsx:
```diff
@@ -15,20 +16,19 @@ const propTypes = {
+  handleSolutionModalHide: () => void;
+  isOpen: boolean;
+  projectTitle: string;
+  solution: string | null | undefined;
```
```suggestion
  solution: string;
```

nhcarrigan replied:
This throws a compiler error in the `show-project-links.tsx` file because the initial state of `solution` is `null`.

nhcarrigan replied:
S'why I proposed the `String(solution)` approach.

hacky, and I'm open to a better way to do it, but it would prevent blocking this PR.

ShaunSHamilton commented on client/src/components/SolutionViewer/ProjectModal.tsx:
```diff
@@ -15,20 +16,19 @@ const propTypes = {
-    solution,
-    handleSolutionModalHide
-  } = props;
+const ProjectModal: React.FC<ProjectModalProps> = ({
+  isOpen,
+  projectTitle,
+  challengeFiles,
+  solution,
+  handleSolutionModalHide
+}: ProjectModalProps) => {
```
```suggestion
const ProjectModal = ({
  isOpen,
  projectTitle,
  challengeFiles,
  solution,
  handleSolutionModalHide
}: ProjectModalProps): JSX.Element => {
```

ShaunSHamilton commented on client/src/components/SolutionViewer/ProjectModal.tsx:
```diff
@@ -45,7 +45,10 @@ const ProjectModal = props => {
       </Modal.Header>
       <Modal.Body>
-        <SolutionViewer challengeFiles={challengeFiles} solution={solution} />
+        <SolutionViewer
+          challengeFiles={challengeFiles}
+          solution={String(solution)}
+        />
```
```suggestion
        <SolutionViewer challengeFiles={challengeFiles} solution={solution} />
```

ShaunSHamilton commented on client/src/components/SolutionViewer/ProjectModal.tsx:
```diff
@@ -15,20 +16,19 @@ const propTypes = {
   //     path: PropTypes.string
   //   })
   // ),
```
I cannot make a suggestion for all of this file.

Please delete this comment - this migration is what the `TODO` is referring to.

ShaunSHamilton reviewed (changes requested):
@ashiskumar-1999 

Thank you, for taking the time to contribute.

You have not addressed all of the previous suggestions. Please make use of the GitHub UI in the `Files changed` tab to 'batch' all of these suggestions. Then, click the `commit suggestions` button to automatically accept the changes.

---

Also, please edit your PR title to follow the recommended format described in the [Contributor Docs](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request)

And be aware that as this PR has been made from your `main` branch, it will require a reset, once this is merged, before any further contributions can be made.

Hope this helps. If you are stuck on anything, or do not understand, please feel free to reach out to us on either [Chat](https://chat.freecodecamp.org/) or the forum.

ashiskumar-1999 commented:
Hey @ShaunSHamilton, I have added the changes in the files. There is no need of adding those changes again.

ashiskumar-1999 reviewed (commented):
Thank you for reviewing the errors.

ashiskumar-1999 committed:
Update ProjectModal.tsx

ashiskumar-1999 committed:
Apply suggestions from code review

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ashiskumar-1999 commented:
Hey @ShaunSHamilton, I don't think that there is any need for changes to those requests because those have been already changed by me. If I won't resolve those conversations then it will throw the error of unresolved conversations. But if I resolve those conversations then the only error came up i.e the "changes requested" error. Can you please tell me how to avoid those conversations?

ShaunSHamilton commented:
> I don't think that there is any need for changes to those requests because those have been already changed by me.

You have not made the suggested changes in the requests.

Currently, one change is still being discussed. As such, it should not be marked as resolved. For the 3rd time, I have marked the conversation as _unresolved_ . Please do not mark a suggestion/comment as resolved, unless you have taken the suggestion.

> If I won't resolve those conversations then it will throw the error of unresolved conversations.

There is no error being thrown. If you commit the suggestion in a review comment, then GitHub will automatically mark the conversation as resolved. Otherwise, it is most agreeable to let the one who made the comment mark the comment as resolved themselves.

This helps avoid the situation of ignored reviews, and duplicate suggestions.

> Can you please tell me how to avoid those conversations?

I do not understand what you are trying to avoid. This is what needs to happen in this PR:
1) A solution to the currently **unresolved** conversation needs to be decided
2) You may help commit the decided change
3) The original reviewer (me, in this case) needs to mark the conversation as resolved
4) The PR gets reviewed again
5) Once all QA checks are complete, a maintainer can merge the PR

Hope this helps

ashiskumar-1999 committed:
Update client/src/components/SolutionViewer/ProjectModal.tsx

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ashiskumar-1999 committed:
Apply suggestions from code review

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ashiskumar-1999 commented:
Yes, I'll do it.

ashiskumar-1999 committed:
Apply suggestions from code review

Co-authored-by: Nicholas Carrigan (he/him) <nhcarrigan@gmail.com>

ashiskumar-1999 committed:
type of solution prop has been changed into "solution?:string;"

ShaunSHamilton committed:
assert solution is not `null` for modal

ShaunSHamilton reviewed (approved):
Thanks, for doing that 👍 

I just pushed a small change asserting the type for the modal-related components.

---

This LGTM
Otherwise, some more changes will be needed once/when the `Certification.js` file is migrated.

naomi-lgbt reviewed (approved):
Looks good to me! Thanks for all your work on this.
